crates/data_components/src/s3_vectors/partition.rs

```diff
@@ -0,0 +1,231 @@
+/*
+Copyright 2024-2025 The Spice.ai OSS Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+     https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+use datafusion::{prelude::Expr, scalar::ScalarValue};
+use snafu::prelude::*;
+use twox_hash::XxHash64;
+
+const HASH_SEED: u64 = 7;
+
+const INDEX_NAME_MAX_LENGTH: usize = 25;
+const COLUMN_NAME_MAX_LENGTH: usize = 25;
+const PARTITION_VALUE_MAX_LENGTH: usize = 5;
+const PARTITION_BY_MAX_LENGTH: usize = 5;
+
+static PARTS_SEPARATOR: &str = "-";
+
+#[derive(Debug, Snafu)]
+pub enum Error {
+    #[snafu(display("Expected 4 parts in index name between hypens but got {num_parts}"))]
+    IncorrectNumPartsInName { num_parts: usize },
+}
+
+#[derive(Debug)]
+pub struct PartitionedIndexName {
+    index_name: String,
+    column_name: String,
+    partition_value_hash: String,
+    partition_by_hash: String,
+}
+
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+pub enum BelongsWith {
+    SameDataset,
+    DifferentDataset,
+    DifferentParitionByExpressions,
+}
+
+impl PartitionedIndexName {
+    pub fn new(
+        index_name: &str,
+        column_name: &str,
+        partition_value: &ScalarValue,
+        partition_by: &[Expr],
+    ) -> Result<Self, Error> {
+        let index_name = truncate(&sanitize(index_name), INDEX_NAME_MAX_LENGTH);
+        let column_name = truncate(&sanitize(column_name), COLUMN_NAME_MAX_LENGTH);
+        let partition_value_hash = truncate(
+            &hash_to_hex(&partition_value.to_string()),
+            PARTITION_VALUE_MAX_LENGTH,
+        );
+        let partition_by_hash = truncate(
+            &hash_to_hex(&format!("{partition_by:?}")),
```

Comment:
Is the debug representataion of an Expression guaranteed to be stable across DF versions? If not, we may want to built our own stable representation so that upgrading Spice won't cause us to lose the association to existing indexes.

Reply:
I'm going to pause and think harder on this because it's really quite the undertaking. Expr variants have many other types inside of them and we would be committing to making a stable representation for every type that appears in every Expr variant. I don't think we want to own this code and deal with the maintenance of it.

Reply:
If we agree SQL is stable then this is unnecessary. It would just be code we have to write and maintain. If SQL wasn't stable I would agree we have to do this as I believed earlier. Is there a good reason not to just use the unparsed SQL?

Reply:
In an ideal world it should be, but the `expr_to_sql` code isn't guaranteed to return the same SQL for a given expression. i.e. this PR very likely would have broken this logic: https://github.com/apache/datafusion/pull/11186

Reply:
Hmm I understand that the SQL is not stable now. That's unfortunate. I'll work on the subset of expressions.